Make libsupport shared always

ddn0 opened the pull request:
```
While debugging an assertion failure in Result.cpp, I noticed that is
possible for a program to have multiple definitions of libsupport. For
example,

  my-lib.so -> libsupport.a
  application -> ... -> libsupport.a
  application -> my-lib.so

While previously this was benign though duplicative; now it can create
more serious issues because katana::Result uses a thread-local buffer to
maintain additional error information. If this buffer can be defined
multiple times, error context information can be accumulated in
different places depending on which version of libsupport a caller
referenced.

Given that, at the end of the day, the purpose of libsupport is to
provide a place for common functionality, it seems simpler to just make
libsupport always a shared library, so there will only be one version of
it for any application.
```

arthurp commented:
I think dropping the python changes from the ISS side will cause problems. There is a definition missing in Galois.pxd (`UserContext.push`) that is [used by _sssp.pyx](https://github.com/ddn0/katana/blob/merge/iss/python/galois/_sssp.pyx#L76). Also the CMake rules for the new files are missing as is the integration of the new libpygalois into the python subproject.

I'm willing to do the merge, but I will need to actually run the merge myself so I have access to the conflicts, so I can fix them. Did you have to fix any other conflicts? Or were the python conflicts the only ones?

gurbinder533 commented:
Yes, we will have to merge python stuff manually. Things have diverged a bit, especially the way we are building and linking libraries. If this is the only conflict, I can work with @arthurp to get this merged properly.

ddn0 commented:
> Did you have to fix any other conflicts? Or were the python conflicts the only ones?

The Python conflicts were the only ones.

I'll leave this in your capable hands.

ddn0 committed:
Make libsupport shared always

While debugging an assertion failure in Result.cpp, I noticed that is
possible for a program to have multiple definitions of libsupport. For
example,

  my-lib.so -> libsupport.a
  application -> ... -> libsupport.a
  application -> my-lib.so

While previously this was benign though duplicative; now it can create
more serious issues because katana::Result uses a thread-local buffer to
maintain additional error information. If this buffer can be defined
multiple times, error context information can be accumulated in
different places depending on which version of libsupport a caller
referenced.

Given that, at the end of the day, the purpose of libsupport is to
provide a place for common functionality, it seems simpler to just make
libsupport always a shared library, so there will only be one version of
it for any application.

arthurp commented:
I think you are missing cmake install rules for `libsupport.so`. That's why the conda packaging failed.

arthurp commented:
> I think you are missing cmake install rules for `libsupport.so`. That's why the conda packaging failed.

This isn't true. I'm not sure what's wrong. :-/

arthurp commented on conda_recipe/meta.yaml:
```diff
@@ -57,7 +57,7 @@ outputs:
       - include/tsuba/**
       - lib/libkatana_galois.so*
-      - lib/libkatana_support.a
+      - lib/libkatana_support.so*
```
Ah yes. That's the issue. Good catch.

arthurp reviewed (approved):
LGTM

scober reviewed (approved):
Thank you!
